service/forward_service.cc

```diff
@@ -0,0 +1,176 @@
+            flogger.error("forward_request: result row count != 1");
+            throw std::runtime_error("aggregation result row count != 1");
+        }
+        if (rows[0].size() != 1) {
+            flogger.error("forward_request: result column count != 1");
+            throw std::runtime_error("aggregation result column count != 1");
+        }
+        auto& result_bytes = rows[0][0];
+        query::forward_result res = { .query_response = result_bytes };
+        return res;
+    });
+}
+
+void forward_service::init_messaging_service() {
+    auto& ms = _messaging;
+    ms.register_forward_request([this](query::forward_request req) -> future<query::forward_result> {
+        return execute(req);
+    });
+}
+
+future<> forward_service::uninit_messaging_service() {
+    return _messaging.unregister_forward_request();
+}
+
+future<query::forward_result> forward_service::dispatch(query::forward_request req) {
+    schema_ptr schema = local_schema_registry().get(req.cmd.schema_version);
+    keyspace& ks = _db.local().find_keyspace(schema->ks_name());
+
+    (void)_shared_token_metadata;
+
+    // next_vnode is used to iterate through all vnodes produced by
+    // query_ranges_to_vnodes_generator
+    auto next_vnode = [
+        generator = query_ranges_to_vnodes_generator(get_token_metadata_ptr(), schema, req.pr)
+    ] () mutable -> std::optional<dht::partition_range> {
+        if (auto vnode = generator(1); !vnode.empty()) {
+            return vnode[0];
+        }
+        return {};
+    };
+
+    // group vnodes by assigned endpoint
+    std::map<netw::messaging_service::msg_addr, dht::partition_range_vector> vnodes_per_addr;
+    while (std::optional<dht::partition_range> vnode = next_vnode()) {
+        inet_address_vector_replica_set live_endpoints = get_live_endpoints(ks, end_token(*vnode));
+        if (live_endpoints.empty()) {
+            throw std::runtime_error("No live endpoint available");
+        }
+
+        auto endpoint_addr = netw::messaging_service::msg_addr{*live_endpoints.begin(), 0};
+        vnodes_per_addr[endpoint_addr].push_back(*vnode);
+    }
+
+    std::optional<query::forward_result> result;
+    // forward request to each endpoint and merge results
+    co_await parallel_for_each(vnodes_per_addr.begin(), vnodes_per_addr.end(),
+        [this, &req, &result] (auto vnodes_with_addr) -> future<> {
+            auto& addr = vnodes_with_addr.first;
+            auto& partition_range = vnodes_with_addr.second;
+            auto req_with_modified_pr = req;
+            req_with_modified_pr.pr = partition_range;
+
+            flogger.debug("dispatching forward_request={} to address={}", req_with_modified_pr, addr);
+            auto partial_result = co_await _messaging.send_forward_request(addr, req_with_modified_pr);
+
+            if (result) {
+                result->merge(partial_result, req.type);
+            } else {
+                result = partial_result;
+            }
+        }
+    );
+
```

Comment:
Please add tracepoints (and verify that executing such a query with cqlsh TRACING ON shows tracing at all three layers).

Reply:
^

Reply:
What do you mean by that? I've added tracepoints in 153d3a8fd17185b5dca8ffa2b01dbc66f057b4fc

Reply:
I didn't see it when I re-reviewed. It is customary to acknowledge a request from a reviewer, or explain why it is incorrect, so the reviewer doesn't have to dig for the response in the code.